File: src/Winwright/Verdicts/Holes.cs
using System.Collections.ObjectModel;
using System.Reflection;

namespace Winwright.Verdicts;

/// <summary>Whose a hole is — which is what decides what the reader does about it.</summary>
public enum Whose
{
    /// <summary>
    /// The desk's. The machine could have been arranged differently: who holds the foreground, what
    /// is standing over a rectangle, whether the shell answers. Nothing to fix in any repository.
    /// </summary>
    Desk,

    /// <summary>
    /// The thing under test's. A stale binary, a page still computing, an application in the wrong
    /// language. A repository to open, and re-running changes nothing.
    /// </summary>
    UnderTest,

    /// <summary>
    /// Neither, because nobody has said. The condition was composed at the throw site, or it is one
    /// this engine declares and <see cref="DeskFacts" /> has never classified — or the hole carries
    /// no condition at all, which is worse than a failure and is counted here rather than rounded
    /// into one of the other two.
    /// </summary>
    Unclassified,
}

/// <summary>
/// How a run's holes divided, with the third bucket kept.
/// </summary>
/// <param name="Desk">Holes about the machine.</param>
/// <param name="UnderTest">Holes about the thing being tested.</param>
/// <param name="Unclassified">Holes nobody has classified, counted rather than assigned.</param>
public sealed record HoleDivision(int Desk, int UnderTest, int Unclassified)
{
    /// <summary>How many holes this divided.</summary>
    public int Total => Desk + UnderTest + Unclassified;

    /// <summary>
    /// The clause a headline carries. One kind reads as <em>all</em>, because "3 the desk's" beside
    /// "3 unchecked" is the same number twice and a reader has to check they are the same number.
    /// </summary>
    public string Sentence()
    {
        if (Total == 0)
            return "none";

        var parts = new List<string>();
        if (Desk > 0)
            parts.Add($"{Desk} {Holes.Worded(Whose.Desk)}");
        if (UnderTest > 0)
            parts.Add($"{UnderTest} {Holes.Worded(Whose.UnderTest)}");
        if (Unclassified > 0)
            parts.Add($"{Unclassified} {Holes.Worded(Whose.Unclassified)}");

        if (parts.Count > 1)
            return string.Join(", ", parts);

        if (Desk > 0)
            return $"all {Holes.Worded(Whose.Desk)}";

        return UnderTest > 0
            ? $"all {Holes.Worded(Whose.UnderTest)}"
            : $"all {Holes.Worded(Whose.Unclassified)}";
    }
}

/// <summary>
/// Whose each hole in a run was.
/// <para>
/// WW192. WW183 wrote this engine's judgement down — which of its conditions are the desk's, and
/// why each one is — and only the suite ever read it. A summary said three assertions never ran and
/// named the absent condition beside each, which is everything except the thing that decides what
/// happens next. Three holes because a foreground was not granted and a window stood over a capture
/// is a machine to clear. Three holes because a binary was stale and a page was still computing is
/// a repository to open. The exit code is 2 either way.
/// </para>
/// <para>
/// The third bucket is not politeness. A condition composed at a throw site is in no catalogue and
/// answers to nothing — WW190 found one of those — and a hole carrying no condition at all is the
/// shape <c>BusyDesk</c> calls worse than a failure. Rounding either into "the desk's" would excuse
/// a defect, and rounding it into "this run's" would send a reader to the wrong repository. Counted,
/// and named in the line, so somebody can go and classify it.
/// </para>
/// </summary>
public static class Holes
{
    /// <summary>How each kind is spelled wherever it is printed.</summary>
    public static string Worded(Whose whose) => whose switch
    {
        Whose.Desk => "the desk's",
        Whose.UnderTest => "this run's",
        _ => "unclassified",
    };

    /// <summary>
    /// Every condition this engine declares, read off its own assembly rather than off a list.
    /// <para>
    /// Two spellings, because the engine has two: a constant named for the precondition it is, and
    /// <c>Named</c> where the reading it belongs to reads better that way. A name nothing here
    /// declares is one somebody composed at the call, and that is the finding rather than an
    /// oversight to be forgiven.
    /// </para>
    /// </summary>
    public static IReadOnlyList<string> Declared { get; } = new ReadOnlyCollection<string>(
        typeof(Precondition).Assembly
            .GetExportedTypes()
            .SelectMany(one => one.GetFields(BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly))
            .Where(one => one.IsLiteral && !one.IsInitOnly && one.FieldType == typeof(string))
            .Where(one => one.Name.Contains("PreconditionName", StringComparison.Ordinal)
                || string.Equals(one.Name, "Named", StringComparison.Ordinal))
            .Select(one => one.GetRawConstantValue() as string ?? "")
            .Where(one => one.Length > 0)
            .Distinct(StringComparer.Ordinal)
            .OrderBy(one => one, StringComparer.Ordinal)
            .ToList());

    /// <summary>
    /// Whose one absent condition is.
    /// <para>
    /// Null is answered rather than refused, and it is worth saying why it cannot arrive from a hole
    /// this engine built: <see cref="AssertionResult.Unchecked" /> refuses one. The parameter is
    /// nullable because <see cref="AssertionResult.Missing" /> is, and a caller holding a result it
    /// did not build gets an answer instead of a throw.
    /// </para>
    /// </summary>
    /// <param name="missing">The condition, or null where nothing named one.</param>
    public static Whose Of(Precondition? missing)
    {
        if (missing is null || string.IsNullOrWhiteSpace(missing.Name))
            return Whose.Unclassified;

        if (DeskFacts.Names(missing.Name))
            return Whose.Desk;

        // Declared and not the desk's is a fact about the thing under test — which is the judgement
        // DeskFacts already makes and states, one call away rather than repeated here.
        return Declared.Contains(missing.Name.Trim(), StringComparer.Ordinal)
            ? Whose.UnderTest
            : Whose.Unclassified;
    }

    /// <summary>Whose one hole is. A result that ran is not a hole and is refused.</summary>
    /// <param name="hole">The assertion that never ran.</param>
    /// <exception cref="ArgumentException">Where it did run, either way.</exception>
    public static Whose Of(AssertionResult hole)
    {
        ArgumentNullException.ThrowIfNull(hole);

        if (hole.Outcome != AssertionOutcome.Unchecked)
        {
            throw new ArgumentException(
                $"'{hole.Name}' {hole.Outcome.ToString().ToLowerInvariant()}, and an assertion that ran "
                    + "is not a hole anybody has to apportion",
                nameof(hole));
        }

        return Of(hole.Missing);
    }

    /// <summary>How a run's holes divide. Anything that ran is passed over rather than refused.</summary>
    /// <param name="results">Every result, or only the holes — either is answered the same way.</param>
    public static HoleDivision Divide(IEnumerable<AssertionResult> results)
    {
        ArgumentNullException.ThrowIfNull(results);

        var desk = 0;
        var underTest = 0;
        var unclassified = 0;

        foreach (var hole in results.Where(one => one.Outcome == AssertionOutcome.Unchecked))
        {
            switch (Of(hole.Missing))
            {
                case Whose.Desk:
                    desk++;
                    break;
                case Whose.UnderTest:
                    underTest++;
                    break;
                default:
                    unclassified++;
                    break;
            }
        }

        return new HoleDivision(desk, underTest, unclassified);
    }
}

File: site/scripts/holes.mjs
// The third verdict, and every condition that earns it. WW491.
//
// An adopter's first surprising run answers `2`. Nothing failed, nothing passed, and the summary
// names an assertion that never ran. That is this project's whole argument arriving at the worst
// possible moment to have to go looking for it — and it was spread over four separate parts of a
// README, none of them titled anything a person would search for.
//
// What the page needs is a list of every condition a hole can name, and whose each one is. Both
// already exist as data in the engine, for the reason WW183 gives: the judgement about a
// condition belongs beside the reading it is about, or it gets made by whoever next has a red
// they cannot explain.
//
// Three sources:
//   src/Winwright/Verdicts/Holes.cs       the three buckets, and what a reader does about each
//   src/Winwright/Verdicts/DeskFacts.cs   the conditions that are the desk's, and why each is
//   src/Winwright/**/*.cs                 every condition this engine declares
//
// That last sweep is the one worth being careful about, because it is a second reading of what
// `Holes.Declared` gets by reflection and Node has no assembly to reflect over. `DeskFactTests`
// holds the two against each other, so a condition declared some third way is a red in the suite
// rather than a row missing here.
import { mkdirSync, readdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { bodyOf, code, constructions, documented, splitTop } from "./csharp.mjs";

const here = dirname(fileURLToPath(import.meta.url));
const siteDir = join(here, "..");
const repoDir = join(siteDir, "..");
const engine = join(repoDir, "src", "Winwright");

const read = (...parts) => readFileSync(join(engine, ...parts), "utf8");

/** Every source of the engine, by the type its file is named for. */
const sources = readdirSync(engine, { recursive: true, withFileTypes: true })
  .filter((one) => one.isFile() && one.name.endsWith(".cs"))
  .filter((one) => !/[\\/](bin|obj)[\\/]/.test(one.parentPath ?? one.path))
  .map((one) => ({ type: one.name.slice(0, -".cs".length), path: join(one.parentPath ?? one.path, one.name) }));

// --- every condition this engine declares ---
// The two spellings `Holes.Declared` selects on: a constant named for the precondition it is, and
// `Named` where the reading it belongs to reads better that way.
// Comments off and strings kept, which is the reading the suite's own sweep makes: a doc comment
// quoting a declaration would otherwise be swept as a second one.
const declared = [
  ...new Set(
    sources.flatMap((one) =>
      [...code(readFileSync(one.path, "utf8")).matchAll(/public const string (?:[A-Za-z]*PreconditionName|Named) = "([^"]+)";/g)]
        .map((m) => m[1]),
    ),
  ),
].sort();

if (declared.length === 0) throw new Error("holes: the engine declares no precondition this can read");

// --- the three buckets ---
// What a reader does about a hole is the bucket's own sentence: the desk's is a machine to clear,
// this run's is a repository to open, and unclassified is somebody's to go and classify.
const kinds = documented(
  bodyOf(read("Verdicts", "Holes.cs"), "public enum Whose", "{", "}", "Holes.cs"),
  (line) => /^([A-Z][A-Za-z]*),$/.exec(line)?.[1],
  "Holes.cs's Whose",
  false,
).map((one) => ({ kind: one.name, means: one.means }));

// --- the conditions that are the desk's ---
// Each entry names its condition through the constant the reading declares, so the name has to be
// followed rather than printed: `Foreground.PreconditionName` is not a sentence anybody meets.
// Indexed by the type that declares each constant, and never by the file it sits in: `Focus.cs`
// declares `FocusReading` beside `Focus`, so a lookup by file name finds neither.
const held = new Map();
for (const one of sources) {
  let declaring = "";
  for (const line of readFileSync(one.path, "utf8").split(/\r?\n/)) {
    const type = /^\s*(?:public|internal)\s+(?:static\s+|sealed\s+|abstract\s+|partial\s+)*(?:class|record|struct|interface)\s+([A-Z][A-Za-z0-9]*)/.exec(line);
    if (type) {
      declaring = type[1];
      continue;
    }

    const constant = /public const string ([A-Za-z]+) = "([^"]*)"/.exec(line);
    if (constant && declaring.length > 0) held.set(`${declaring}.${constant[1]}`, constant[2]);
  }
}

function conditionOf(expression, where) {
  const named = /^([A-Z][A-Za-z]*)\.([A-Za-z]+)$/.exec(expression.trim());
  if (!named) throw new Error(`holes: ${where} names ${expression.trim()}, which is no constant this can follow`);

  const value = held.get(named[0]);
  if (value === undefined) throw new Error(`holes: nothing in the engine declares ${named[0]} for ${where} to name`);
  return value;
}

/** A quoted literal, or several concatenated across the lines a long sentence wraps over. */
function sentence(expression, where) {
  return splitTop(expression, "+")
    .map((part) => {
      const literal = /^"([\s\S]*)"$/.exec(part.trim());
      if (!literal) throw new Error(`holes: ${where} is not a sentence this can read: ${part.trim()}`);
      return literal[1].replace(/\\"/g, '"');
    })
    .join("");
}

const desk = constructions(
  bodyOf(code(read("Verdicts", "DeskFacts.cs")), "public static IReadOnlyList<DeskFact> Known", "[", "]", "DeskFacts.cs"),
).map((one) => {
  const args = splitTop(one, ",");
  if (args.length !== 2) {
    throw new Error(`holes: a DeskFact has ${args.length} arguments and it takes 2: ${one.trim().slice(0, 60)}`);
  }

  const named = conditionOf(args[0], "a DeskFact");
  return { named, because: sentence(args[1], `the reason for '${named}'`) };
});

if (desk.length === 0) throw new Error("holes: DeskFacts.Known calls nothing the desk's");

for (const one of desk) {
  // A desk fact naming a condition nothing declares excuses nothing, which is the finding WW183
  // is about. The suite says so too; saying it here keeps the page from publishing the orphan.
  if (!declared.includes(one.named)) {
    throw new Error(`holes: the desk fact '${one.named}' names a condition this engine does not declare`);
  }
}

const deskNames = new Set(desk.map((one) => one.named));
const underTest = declared.filter((one) => !deskNames.has(one));

// --- the precedence ---
// Not the enum's order, and the difference matters: taking the largest member value would rank a
// hole above a failure. Read off the fold a suite verdict makes rather than typed beside it.
const fold = bodyOf(
  code(readFileSync(join(repoDir, "src", "Winwright", "Scenarios", "Suite.cs"), "utf8")),
  "internal SuiteVerdict(",
  "{",
  "}",
  "Suite.cs",
);

const precedence = [...fold.matchAll(/readings\.Contains\(RunOutcome\.([A-Za-z]+)\)/g)].map((m) => m[1]);
if (precedence.length === 0) throw new Error("holes: the suite verdict no longer folds its readings by outcome");
if (!/readings\.Count > 0 \? RunOutcome\.([A-Za-z]+)/.test(fold)) {
  throw new Error("holes: the suite verdict no longer says what a run of no readings is");
}

precedence.push(/readings\.Count > 0 \? RunOutcome\.([A-Za-z]+)/.exec(fold)[1]);

const holes = {
  kinds,
  desk,
  underTest,
  precedence,
  // What a fold with nothing in it answers, which is the one a reader is most likely to disbelieve.
  empty: /readings\.Count > 0 \? RunOutcome\.[A-Za-z]+\s*:\s*RunOutcome\.([A-Za-z]+)/.exec(fold)?.[1] ?? "",
  counts: { conditions: declared.length, desk: desk.length, underTest: underTest.length },
};

if (!holes.empty) throw new Error("holes: nothing says what a run of no readings answers");

const out = join(siteDir, "docs", "src", "data", "holes.generated.json");
mkdirSync(dirname(out), { recursive: true });
writeFileSync(out, `${JSON.stringify(holes, null, 2)}\n`);

console.log(
  `holes: ${holes.counts.conditions} condition(s), ${holes.counts.desk} the desk's,`
    + ` ${holes.counts.underTest} this run's; ${precedence.join(" > ")}, and a run of none is ${holes.empty}`
    + " -> docs/src/data/holes.generated.json",
);
